File: packages/card/apf/renderer.py
"""APF Renderer - Convert JSON CardSpec to APF v2.1 HTML.

This module provides deterministic conversion from structured JSON
card specifications to APF HTML format, eliminating the risk of
truncated or malformed output from LLMs.
"""

from __future__ import annotations

import html
import json
from typing import Any


class APFRenderer:
    """Convert JSON CardSpec to APF v2.1 HTML.

    This renderer produces deterministic, well-formed APF output
    from structured card specifications. All required sentinels
    and markers are guaranteed to be present.

    The ``spec`` parameter uses duck typing (any object with the expected
    attributes) so that this package does not depend on the generation
    package where ``CardSpec`` is defined.
    """

    PROMPT_VERSION = "apf-v2.1"

    def render(self, spec: Any) -> str:
        """Render a CardSpec to APF v2.1 HTML."""
        parts = [
            f"<!-- PROMPT_VERSION: {self.PROMPT_VERSION} -->",
            "<!-- BEGIN_CARDS -->",
            "",
            self._render_card_header(spec),
            "",
            self._render_title(spec.front.title),
            "",
            self._render_key_point(spec.front),
            "",
            self._render_key_point_notes(spec.front.key_point_notes),
            "",
            self._render_other_notes(spec.front.other_notes),
            "",
            self._render_extra(spec.front.extra),
            "",
            self._render_manifest(spec),
            "<!-- END_CARDS -->",
            "END_OF_CARDS",
        ]
        return "\n".join(parts)

    def render_batch(self, specs: list[Any]) -> str:
        """Render multiple cards separated by card markers."""
        if not specs:
            return ""
        if len(specs) == 1:
            return self.render(specs[0])

        rendered_cards = [self.render(spec) for spec in specs]
        return "\n\n<!-- CARD_SEPARATOR -->\n\n".join(rendered_cards)

    def _render_card_header(self, spec: Any) -> str:
        """Render the card header comment."""
        tags_str = " ".join(spec.tags) if spec.tags else ""
        return (
            f"<!-- Card {spec.card_index} | slug: {spec.slug} | "
            f"CardType: {spec.card_type} | Tags: {tags_str} -->"
        )

    def _render_title(self, title: str) -> str:
        """Render the title section."""
        return f"<!-- Title -->\n{title}"

    def _render_key_point(self, section: Any) -> str:
        """Render the key point (code block) section."""
        header = "<!-- Key point (code block / image) -->"

        if not section.key_point_code:
            return header

        lang = section.key_point_code_lang or "plaintext"
        escaped_code = html.escape(section.key_point_code)

        return (
            f'{header}\n<pre><code class="language-{lang}">'
            f"{escaped_code}</code></pre>"
        )

    def _render_key_point_notes(self, notes: list[str]) -> str:
        """Render the key point notes section."""
        header = "<!-- Key point notes -->"

        if not notes:
            return f"{header}\n<ul></ul>"

        items = [f"<li>{note}</li>" for note in notes]
        items_str = "\n".join(items)
        return f"{header}\n<ul>\n{items_str}\n</ul>"

    def _render_other_notes(self, other_notes: str) -> str:
        """Render the other notes section."""
        header = "<!-- Other notes -->"

        if not other_notes:
            return header

        return f"{header}\n{other_notes}"

    def _render_extra(self, extra: str) -> str:
        """Render the extra section."""
        header = "<!-- Extra -->"

        if not extra:
            return header

        return f"{header}\n{extra}"

    def _render_manifest(self, spec: Any) -> str:
        """Render the manifest comment."""
        manifest: dict[str, Any] = {
            "slug": spec.slug,
            "slug_base": spec.slug_base or spec.slug.rsplit("-", 2)[0],
            "lang": spec.lang,
            "type": spec.card_type,
            "tags": spec.tags,
            "guid": spec.guid,
        }

        if spec.source_path:
            manifest["source_path"] = spec.source_path
        if spec.source_anchor:
            manifest["source_anchor"] = spec.source_anchor

        manifest_json = json.dumps(
            manifest, ensure_ascii=False, separators=(",", ":")
        )
        return f"<!-- manifest: {manifest_json} -->"


class APFSentinelValidator:
    """Validate APF structure has all required sentinels."""

    REQUIRED_SENTINELS = (
        "<!-- PROMPT_VERSION:",
        "<!-- BEGIN_CARDS -->",
        "<!-- Card ",
        "<!-- Title -->",
        "<!-- Key point",
        "<!-- manifest:",
        "<!-- END_CARDS -->",
    )

    def validate(self, apf_html: str) -> list[str]:
        """Return list of missing sentinel markers (empty if all present)."""
        if not apf_html:
            return list(self.REQUIRED_SENTINELS)

        return [s for s in self.REQUIRED_SENTINELS if s not in apf_html]

    def is_valid(self, apf_html: str) -> bool:
        """Return True if all required sentinels are present."""
        return len(self.validate(apf_html)) == 0

File: crates/generator/src/apf/renderer.rs
use serde::{Deserialize, Serialize};

pub const PROMPT_VERSION: &str = "apf-v2.1";

/// Input spec for rendering a card to APF HTML.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CardSpec {
    pub card_index: u32,
    pub slug: String,
    pub slug_base: Option<String>,
    pub lang: String,
    pub card_type: String,
    pub tags: Vec<String>,
    pub guid: String,
    pub source_path: Option<String>,
    pub source_anchor: Option<String>,
    pub title: String,
    pub key_point_code: Option<String>,
    pub key_point_code_lang: Option<String>,
    pub key_point_notes: Vec<String>,
    pub other_notes: Option<String>,
    pub extra: Option<String>,
}

fn escape_html(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
}

fn build_manifest(spec: &CardSpec) -> String {
    let slug_base = spec
        .slug_base
        .clone()
        .unwrap_or_else(|| match spec.slug.rfind('-') {
            Some(pos) => spec.slug[..pos].to_string(),
            None => spec.slug.clone(),
        });

    let tags_json: Vec<String> = spec.tags.iter().map(|t| format!("\"{t}\"")).collect();

    let mut parts = vec![
        format!("\"slug\":\"{}\"", spec.slug),
        format!("\"slug_base\":\"{}\"", slug_base),
        format!("\"lang\":\"{}\"", spec.lang),
        format!("\"type\":\"{}\"", spec.card_type),
        format!("\"tags\":[{}]", tags_json.join(",")),
    ];

    if let Some(ref path) = spec.source_path {
        parts.push(format!("\"source_path\":\"{path}\""));
    }
    if let Some(ref anchor) = spec.source_anchor {
        parts.push(format!("\"source_anchor\":\"{anchor}\""));
    }

    format!("{{{}}}", parts.join(","))
}

/// Render a single card spec to APF HTML.
pub fn render(spec: &CardSpec) -> String {
    let tags_str = spec.tags.join(" ");
    let mut out = String::new();

    // Header sentinels
    out.push_str(&format!("<!-- PROMPT_VERSION: {PROMPT_VERSION} -->\n"));
    out.push_str("<!-- BEGIN_CARDS -->\n");
    out.push_str(&format!(
        "<!-- Card {} | slug: {} | CardType: {} | Tags: {} -->\n",
        spec.card_index, spec.slug, spec.card_type, tags_str
    ));

    // Title
    out.push_str(&format!("<!-- Title -->\n{}\n", spec.title));

    // Key point code block
    out.push_str("<!-- Key point (code block / image) -->\n");
    if let Some(ref code) = spec.key_point_code {
        let lang = spec.key_point_code_lang.as_deref().unwrap_or("plaintext");
        out.push_str(&format!(
            "<pre><code class=\"language-{}\">{}</code></pre>\n",
            lang,
            escape_html(code)
        ));
    }

    // Key point notes
    out.push_str("<!-- Key point notes -->\n");
    if spec.key_point_notes.is_empty() {
        out.push_str("<ul></ul>\n");
    } else {
        out.push_str("<ul>");
        for note in &spec.key_point_notes {
            out.push_str(&format!("<li>{note}</li>"));
        }
        out.push_str("</ul>\n");
    }

    // Other notes
    out.push_str("<!-- Other notes -->\n");
    if let Some(ref notes) = spec.other_notes {
        out.push_str(&format!("{notes}\n"));
    }

    // Extra
    out.push_str("<!-- Extra -->\n");
    if let Some(ref extra) = spec.extra {
        out.push_str(&format!("{extra}\n"));
    }

    // Manifest
    out.push_str(&format!("<!-- manifest:{} -->\n", build_manifest(spec)));

    // Footer sentinels
    out.push_str("<!-- END_CARDS -->\n");
    out.push_str("END_OF_CARDS");

    out
}

/// Render multiple card specs, separated by `<!-- CARD_SEPARATOR -->`.
pub fn render_batch(specs: &[CardSpec]) -> String {
    if specs.is_empty() {
        return String::new();
    }

    specs
        .iter()
        .map(render)
        .collect::<Vec<_>>()
        .join("\n<!-- CARD_SEPARATOR -->\n")
}
